### ISSUE TITLE
How can I use ILogger that is registered with JobHost outside WebJobs functions?

### ISSUE THREAD

**Issue by wind-rider** [Azure/azure-webjobs-sdk#1229]
I registered logging in my WebJob according to the [Application Insights](https://github.com/Azure/azure-webjobs-sdk/wiki/Application-Insights-Integration) page:

```csharp
var config = new JobHostConfiguration();

// If this variable exists, build up a LoggerFactory with ApplicationInsights and a Console Logger
string instrumentationKey = Environment.GetEnvironmentVariable("APPINSIGHTS_INSTRUMENTATIONKEY");
if (!string.IsNullOrEmpty(instrumentationKey))
{
    // Wire up with default filters; Filtering will be explained later.
    config.LoggerFactory = new LoggerFactory()
        .AddApplicationInsights(instrumentationKey, null)
        .AddConsole();

    config.Tracing.ConsoleLevel = TraceLevel.Off;
}
```

I see that these loggers can be used in the triggered functions, but it is unclear to me how to access the ILogger implementations in the main WebJob process (like [how it works with dependency injection in ASP.Net Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging).

I'd like to use it with dependency injection, like Autofac. How should I register these loggers for them to be available in the DI container?

I see that one can call loggerFactory.CreateLogger<T>() but it is unclear for me how to use it in dependency injection then.

Please don't close it saying that it is an autofac issue - it is not; I want to understand how to use dependency injection properly in the main part of WebJobs, and how to use the LoggerFactory and Application Insights properly..

**Comment by brettsam** [comment]
Can you explain what you're looking to do with Autofac? Where do you want to make the LoggerFactory available?

**Comment by wind-rider** [comment]
@brettsam Autofac is a dependency injection framework; but more in general, I want to use dependency injection with WebJobs. I see that there is:

```csharp
public void AddService<TService>(TService serviceInstance);
public void AddService(Type serviceType, object serviceInstance);
public TService GetService<TService>();
```
but this does not give the full dependency injection functionality as ASP.Net Core does. However using Autofac and by [implementing IJobActivator](https://stackoverflow.com/a/30454556/1114918) it is possible to get dependency injection in Webjobs:

```csharp
public class AutofacJobActivator : IJobActivator
    {
        private readonly IContainer _container;

        public AutofacJobActivator(IContainer container)
        {
            _container = container;
        }

        public T CreateInstance<T>()
        {
            return _container.Resolve<T>();
        }
    }
```

It would be nice if this would be more aligned with ASP.Net Core.

Furthermore it is unclear to me how to get hold of a logger instance that is registered with the JobHost (like in https://github.com/Azure/azure-webjobs-sdk/wiki/Application-Insights-Integration) using dependency injection outside the "WebJob triggered functions".

**Comment by brettsam** [comment]
Yes, I agree with you and in fact we're in the early stages of some work that will improve this and align it better going forward. A lot of that work is happening in the Script/Functions repo, but I'd like to see what we can improve here as well. 

As for getting the `ILoggerFactory` instance -- you can find it on `JobHostConfiguration.LoggerFactory`. That's why I was curious where you wanted to access it from -- if you have access to that config, then you can grab it. Do you have any sample code for what you're trying to do?

**Comment by XristophD** [comment]
Is there any documentation around using `JobHostConfiguration.LoggerFactory` and accessing the logger within Functions during e.g. `ProcessQueueMessage`?

What I understand is we can now add a LoggerFactory to the JobHost to use this logger within the execution of the WebJob-Function. But I am currently not sure how to grab the logger within e.g. `Functions.ProcessQueueMessage`

**Comment by brettsam** [comment]
From within a function you can use a parameter of type `ILogger` and you'll be given a logger created from the logger factory.
